File: packages/go/internal/seguranca/ataques/ataque_chaves_degeneradas.go
package ataques

import (
	"bytes"
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// OpcoesChavesDegeneradas configura o AtaqueChavesDegeneradas.
type OpcoesChavesDegeneradas struct {
	TamanhoBloco int
}

// AtaqueChavesDegeneradas testa chaves especiais (zero, 0xFF, alternadas,
// baixa entropia) com IV zerado, isolando a contribuição da chave.
type AtaqueChavesDegeneradas struct {
	tamanhoBloco int
}

// NovoAtaqueChavesDegeneradas cria o ataque. Sem argumento, usa
// tamanhoBloco = 32.
func NovoAtaqueChavesDegeneradas(opcoes ...OpcoesChavesDegeneradas) *AtaqueChavesDegeneradas {
	o := OpcoesChavesDegeneradas{TamanhoBloco: 32}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueChavesDegeneradas{tamanhoBloco: o.TamanhoBloco}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueChavesDegeneradas) Nome() string {
	return "Chaves degeneradas (zero, 0xFF, alternada, baixa entropia)"
}

// Executar roda o ataque.
func (a *AtaqueChavesDegeneradas) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	tamChave := len(alvo.ChaveDeTeste())
	ivZero := make([]byte, alvo.TamanhoIv())

	seqCrescente := make([]byte, 512)
	for i := 0; i < 512; i++ {
		seqCrescente[i] = byte(i % 256)
	}

	type caso struct {
		nome  string
		chave []byte
	}
	casos := []caso{
		{nome: "zero", chave: make([]byte, tamChave)},
		{nome: "0xFF", chave: bytes.Repeat([]byte{0xff}, tamChave)},
		{nome: "alternada 0xAA", chave: bytes.Repeat([]byte{0xaa}, tamChave)},
		{nome: "alternada 0x55", chave: bytes.Repeat([]byte{0x55}, tamChave)},
		{nome: "repetida \"A\"", chave: bytes.Repeat([]byte{0x41}, tamChave)},
		{nome: "crescente", chave: seqCrescente[:tamChave]},
	}

	type problema struct {
		nome        string
		periodico   bool
		bytesUnicos int
	}
	problemas := []problema{}
	for _, c := range casos {
		ks := alvo.GerarKeystreamBruto(c.chave, ivZero, "enc", a.tamanhoBloco)

		metade := a.tamanhoBloco / 2
		periodico := bytes.Equal(ks[:metade], ks[metade:metade+metade])
		unicos := map[byte]bool{}
		for _, b := range ks {
			unicos[b] = true
		}
		bytesUnicos := len(unicos)

		if periodico || float64(bytesUnicos) < float64(a.tamanhoBloco)*0.5 {
			problemas = append(problemas, problema{nome: c.nome, periodico: periodico, bytesUnicos: bytesUnicos})
		}
	}

	if len(problemas) > 0 {
		partes := make([]string, len(problemas))
		for i, p := range problemas {
			simNao := "não"
			if p.periodico {
				simNao = "sim"
			}
			partes[i] = fmt.Sprintf("%s (periódico=%s, bytes únicos=%d/%d)", p.nome, simNao, p.bytesUnicos, a.tamanhoBloco)
		}
		dados := map[string]any{}
		for _, p := range problemas {
			dados[p.nome] = map[string]any{"periodico": p.periodico, "bytes_unicos": p.bytesUnicos}
		}
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevAlta,
			Detalhes:   strings.Join(partes, "; "),
			Dados:      dados,
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Nenhuma das %d chaves degeneradas testadas produziu saída anômala", len(casos)),
	}, nil
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueChavesDegeneradas.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Análogo ao AtaqueIVsDegenerados, mas para a CHAVE. Chaves especiais
 * (tudo zero, tudo 0xFF, padrões alternados, baixa entropia) não podem
 * produzir keystream degenerado - e aqui testamos no pior cenário, com IV
 * zerado junto, pra isolar a contribuição da chave.
 */
public class AtaqueChavesDegeneradas implements AtaqueInterface {
    private final int tamanhoBloco;

    public AtaqueChavesDegeneradas() {
        this(32);
    }

    public AtaqueChavesDegeneradas(int tamanhoBloco) {
        this.tamanhoBloco = tamanhoBloco;
    }

    @Override
    public String nome() {
        return "Chaves degeneradas (zero, 0xFF, alternada, baixa entropia)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        int tamChave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8).length;
        byte[] ivZero = new byte[alvo.tamanhoIv()];

        byte[] primeiro = alvo.gerarKeystreamBruto(
                alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8), ivZero, "enc", this.tamanhoBloco);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        byte[] seqCrescente = new byte[512];
        for (int i = 0; i < 512; i++) {
            seqCrescente[i] = (byte) (i % 256);
        }

        Map<String, byte[]> casos = new LinkedHashMap<>();
        casos.put("zero", new byte[tamChave]);
        byte[] ff = new byte[tamChave];
        Arrays.fill(ff, (byte) 0xFF);
        casos.put("0xFF", ff);
        byte[] aa = new byte[tamChave];
        Arrays.fill(aa, (byte) 0xAA);
        casos.put("alternada 0xAA", aa);
        byte[] c55 = new byte[tamChave];
        Arrays.fill(c55, (byte) 0x55);
        casos.put("alternada 0x55", c55);
        byte[] repA = new byte[tamChave];
        Arrays.fill(repA, (byte) 0x41);
        casos.put("repetida \"A\"", repA);
        casos.put("crescente", Arrays.copyOf(seqCrescente, tamChave));

        Map<String, Map<String, Object>> problemas = new LinkedHashMap<>();
        for (Map.Entry<String, byte[]> e : casos.entrySet()) {
            byte[] chave = e.getValue();
            byte[] ks = alvo.gerarKeystreamBruto(chave, ivZero, "enc", this.tamanhoBloco);

            int metade = this.tamanhoBloco / 2;
            boolean periodico = Arrays.equals(
                    Arrays.copyOfRange(ks, 0, metade), Arrays.copyOfRange(ks, metade, metade + metade));
            Set<Integer> unicos = new HashSet<>();
            for (byte b : ks) {
                unicos.add(b & 0xFF);
            }
            int bytesUnicos = unicos.size();

            if (periodico || bytesUnicos < this.tamanhoBloco * 0.5) {
                Map<String, Object> info = new HashMap<>();
                info.put("periodico", periodico);
                info.put("bytes_unicos", bytesUnicos);
                problemas.put(e.getKey(), info);
            }
        }

        if (!problemas.isEmpty()) {
            List<String> detalhes = new ArrayList<>();
            for (Map.Entry<String, Map<String, Object>> e : problemas.entrySet()) {
                Map<String, Object> info = e.getValue();
                detalhes.add(e.getKey() + " (periódico=" + (((Boolean) info.get("periodico")) ? "sim" : "não")
                        + ", bytes únicos=" + info.get("bytes_unicos") + "/" + this.tamanhoBloco + ")");
            }
            return new ResultadoAtaque(
                    nome(), true, Severidade.ALTA, String.join("; ", detalhes), new HashMap<>(problemas));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                "Nenhuma das " + casos.size() + " chaves degeneradas testadas produziu saída anômala");
    }
}
